src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java

```diff
@@ -61,34 +96,177 @@ private void combineShardScores(final ScoreCombinationTechnique scoreCombination
-        return sortedDocsIds;
+        Map<Integer, Object[]> docIdSortFieldMap = new HashMap<>();
+        List<TopDocs> topFieldDocs = compoundTopDocs.getTopDocs();
+        final boolean isSortByScore = checkIfSortOrderByScore(sort);
+        for (TopDocs topDocs : topFieldDocs) {
+            for (ScoreDoc scoreDoc : topDocs.scoreDocs) {
+                FieldDoc fieldDoc = (FieldDoc) scoreDoc;
+
+                if (docIdSortFieldMap.get(fieldDoc.doc) == null) {
+                    // If sort by score then replace sort field value with normalized score.
+                    if (isSortByScore) {
+                        docIdSortFieldMap.put(fieldDoc.doc, new Object[] { combinedNormalizedScoresByDocId.get(fieldDoc.doc) });
+                    } else {
+                        docIdSortFieldMap.put(fieldDoc.doc, fieldDoc.fields);
+                    }
+                }
+            }
+        }
+        return docIdSortFieldMap;
+    }
+
+    private Collection<Integer> getSortedDocIds(
+        final Map<Integer, Float> combinedNormalizedScoresByDocId,
+        final List<TopFieldDocs> topFieldDocs,
+        final Sort sort
+    ) {
+        // we're merging docs with normalized and combined scores. we need to have only maxHits results
+        if (Objects.isNull(sort)) {
+            List<Integer> sortedDocsIds = new ArrayList<>(combinedNormalizedScoresByDocId.keySet());
+            sortedDocsIds.sort((a, b) -> Float.compare(combinedNormalizedScoresByDocId.get(b), combinedNormalizedScoresByDocId.get(a)));
+            return sortedDocsIds;
+        } else {
+            if (Objects.isNull(topFieldDocs)) {
+                throw new IllegalArgumentException("topFieldDocs cannot be null when sorting is enabled.");
+            }
+            int topN = 0;
+            for (TopFieldDocs topFieldDoc : topFieldDocs) {
+                topN += topFieldDoc.scoreDocs.length;
+            }
+
+            // Merge the sorted results of individual queries to form a one final result per shard which is sorted.
+            // Input
+            // < 0, 0.7, shardId, [90]> //Query 1` result scoreDoc
+            // < 1, 0.7, shardId, [70]> //Query 1 result scoreDoc
+            // < 2, 0.3, shardId, [100]> //Query 2 result scoreDoc
+            // < 1, 0.3, shardId, [70]> //Query 2 result scoreDoc
+
+            // Output
+            // < 2, 0.3, shardId, [100]>
+            // < 0, 0.7, shardId, [90]>
+            // < 1, 0.7, shardId, [70]>
+            // < 1, 0.3, shardId, [70]>
+            final TopDocs sortedTopDocs = TopDocs.merge(sort, 0, topN, topFieldDocs.toArray(new TopFieldDocs[0]), SORTING_TIE_BREAKER);
+
+            // Remove duplicates from the sorted top docs.
+            Set<Integer> uniqueDocIds = new LinkedHashSet<>();
+            for (ScoreDoc scoreDoc : sortedTopDocs.scoreDocs) {
+                uniqueDocIds.add(scoreDoc.doc);
+            }
+            return uniqueDocIds;
```

Comment:
why not convert to List<Integer> to keep the interface return type less generic than Collection<Integer> ?

Reply:
I did the same before but @martin-gaievski asked me to go for generic like Collection. The reason behind that is Converting to list would be extra time consuming operation and as search is hotspot we kept Collection over here to let not impact performance. For more details check this comment out https://github.com/opensearch-project/neural-search/pull/802#discussion_r1657658609